### PR TITLE
[JENKINS-21224] Utility `Listeners.notify`

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-21224](https://issues.jenkins-ci.org/browse/JENKINS-21224). Inspired by [this thread](https://groups.google.com/g/jenkinsci-dev/c/TexBP8-nE30/m/ajREF0JSBgAJ).

### Proposed changelog entries

* Introduced `Listeners.notify` utility method to encourage safe listener notification.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/slaves/SlaveComputer.java
Comment (jglick) [inline core/src/main/java/hudson/slaves/SlaveComputer.java:899]: _Possibly_ solving the problem reported by @rockseisen.

File: core/src/main/java/hudson/model/listeners/ItemListener.java
Comment (jglick) [inline core/src/main/java/hudson/model/listeners/ItemListener.java:174]: Just one example of a listener converted to the new API, in addition to `ComputerListener`. There are a lot more that could be done.

File: core/src/main/java/hudson/model/listeners/ItemListener.java
Comment (jglick) [inline core/src/main/java/hudson/model/listeners/ItemListener.java:187]: Deleting null check introduced in #4458 for no reason I can follow. `ExtensionList.lookup` should never return null entries.

**Review by timja — APPROVED** [review]
Very nice

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: core/src/main/java/hudson/model/listeners/ItemListener.java
Comment (basil) [inline core/src/main/java/hudson/model/listeners/ItemListener.java:201]: With this change, the following tests start failing:

- `com.cloudbees.hudson.plugins.folder.properties.AuthorizationMatrixPropertyTest#ensureCreatorHasPermissions`
- `hudson.security.ProjectMatrixAuthorizationStrategyTest#ensureCreatorHasPermissions`
- `hudson.security.ProjectMatrixAuthorizationStrategyTest#ensureNoJobPropertyDuplication`
- `org.jenkinsci.plugins.matrixauth.AuthorizationMatrixNodePropertyTest#ensureCreatorHasPermissions`

Without this change, the tests still pass.

**Pull request merged**
